cloudinit/sources/DataSourceAzure.py

```diff
@@ -660,10 +691,249 @@ def setup(self, is_new_instance):
             LOG.debug("negotiating already done for %s",
                       self.get_instance_id())
 
+    def _wait_for_nic_detach(self, nl_sock):
+        """Use the netlink socket provided to wait for nic detach event.
+           NOTE: The function doesn't close the socket. The caller owns closing
+           the socket and disposing it safely.
+        """
+        try:
+            ifname = None
+
+            # Preprovisioned VM will only have one NIC, and it gets
+            # detached immediately after deployment.
+            ifname = netlink.wait_for_nic_detach_event(nl_sock)
+            if ifname is None:
+                msg = "Preprovisioned nic not detached as expected. " +\
+                    "Proceeding without failing."
+                report_diagnostic_event(msg, logger_func=LOG.warning)
+            else:
+                LOG.info("The preprovisioned nic %s is detached.", ifname)
+            path = REPROVISION_NIC_DETACHED_MARKER_FILE
+            LOG.info("Creating a marker file for nic detached: %s", path)
+            util.write_file(path, "{pid}: {time}\n".format(
+                pid=os.getpid(), time=time()))
+        except AssertionError as error:
+            report_diagnostic_event(error, logger_func=LOG.error)
+            raise
+
+    @azure_ds_telemetry_reporter
+    def wait_for_link_up(self, ifname):
+        """In cases where the link state is still showing down after a nic is
+           hot-attached, we can attempt to bring it up by forcing the hv_netvsc
+           drivers to query the link state by unbinding and then binding the
+           device. This function attempts infinitely until the link is up,
+           because we cannot proceed further until we have a stable link."""
+
+        if self.distro.networking.try_set_link_up(ifname):
+            LOG.info("The link %s is already up.", ifname)
+            return
+
+        LOG.info("Attempting to bring %s up", ifname)
+
+        attempts = 0
+        while True:
+
+            LOG.info("Unbinding and binding the interface %s", ifname)
+            devicename = net.read_sys_net(ifname,
+                                          'device/device_id').strip('{}')
+            util.write_file('/sys/bus/vmbus/drivers/hv_netvsc/unbind',
+                            devicename)
+            util.write_file('/sys/bus/vmbus/drivers/hv_netvsc/bind',
+                            devicename)
+
+            attempts = attempts + 1
+            if self.distro.networking.try_set_link_up(ifname):
+                msg = "The link %s is up after %s attempts" % (ifname,
+                                                               attempts)
+                report_diagnostic_event(msg, logger_func=LOG.info)
+                return
+
+            LOG.info("Link is not up despite attempts to bring it up. "
+                     "Sleeping for 1s and trying again.")
+            sleep(1)
```

Comment:
For every 10 attempts, please log using `report_diagnostic_event`.
```suggestion
            sleep_duration = 1
            msg = ('Link is not up despite %d attempts to bring it up '
                'with %d seconds sleep between each attempt.'
                % (attempts, sleep_duration))
            if attempts % 10 == 0:
                report_diagnostic_event(msg, logger=LOG.info)
            else:
                LOG.info(msg)
            sleep(sleep_duration)
```